cmake/Sanitizers.cmake

```diff
@@ -0,0 +1,53 @@
+# Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#  http://aws.amazon.com/apache2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+include(CheckCCompilerFlag)
+
+function(aws_add_sanitizers target)
+    if(CMAKE_BUILD_TYPE STREQUAL "" OR CMAKE_BUILD_TYPE MATCHES Debug)
+        check_c_compiler_flag(-fsanitize= HAS_SANITIZERS)
+        if(HAS_SANITIZERS)
+            if(ARGN)
+                set(SANITIZERS ${ARGN})
+            else()
+                set(SANITIZERS "address;undefined")
+            endif()
+
+            foreach (sanitizer IN LISTS SANITIZERS)
+
+                # When testing for libfuzzer, if attempting to link there will be 2 mains
+                if(${sanitizer} STREQUAL "fuzzer")
+                    set(sanitizer_test_flag -fsanitize=fuzzer-no-link)
+                else()
+                    set(sanitizer_test_flag -fsanitize=${sanitizer})
+                endif()
+
+                set(sanitizer_variable HAS_SANITIZER_${sanitizer})
+                string(MAKE_C_IDENTIFIER ${sanitizer_variable} sanitizer_variable)
```

Comment:
I don't quite get the reason for this. Can you explain it via comments?

Reply:
Sure!